### ISSUE TITLE
Add built-in button / map & character designer

### ISSUE THREAD

**Issue by bambier** [pygame/pygame#4268]
**Description**
Please add the build-in button with hover, click, drag & etc effect/events to pygame.

There's lot of tutorials on the internet to show how to create buttons in pygame, but none of them were added click effect on button. We need a few milliseconds of sleep before calling the function that should be called after the clicking button.
There's implementation of this feature in other similar frameworks or libraries such as Panda3D maybe inspiring from them is a good approach to fix this problem.

Also adding build-in map & character or tool designer could be useful.

**Comment by IGR2020** [comment]
Hey you can use this basic button class and open hover (while Button.clicked()) which the button image to a on hover image, to get when the button is clicked you can call button.clicked() method under pg.mouseup event.
```
class Button(pygame.Rect):
    def __init__(self, pos, image, scale=1, *args):
        x, y = pos
        width, height = image.get_width() * scale, image.get_height() * scale
        super().__init__(x, y, width, height)
        self.image = pygame.transform.scale(image, (width, height))
        if len(args) == 1:
            self.info = args[0]
        else:
            self.info = args

    def clicked(self):
        pos = pygame.mouse.get_pos()
        if self.collidepoint(pos):
            return True
        return False

    def display(self, win):
        win.blit(self.image, self)
```

**Comment by bambier** [comment]
Thank you for your advice, but I have been using this before. It doesn't work well in every circumstance.

**Comment by IGR2020** [comment]
yes, that is true, effect and particles are very hard to do, most of all zoom features and in built possibilities are not the best.

**Comment by oskarmeyenburg** [comment]
@bambier I think your answer summarizes the problem pretty well: There is no solution for every case.
Some may prefer different images for the normal, hover and pressed state, or even an entire animation of multiple images per state. Others want the button to move when being clicked, make a sound and spawn particles. And again others don't want to draw an image for the button and just use a normal coloured rectangle as the button. 
A built-in button class would either have major performance drawbacks or leave a lot people unhappy because of missing features.

**Comment by bambier** [comment]
Whoever walks may be lost. Who doesn't, will not reach to the distinction ever.

It will not help! Because of that, someone may be unhappy we shouldn't grow? What kind of logic is this 😐😕

If so, we had to add all of those features to make all happy, but at this time, I need a button that can pressed hoverd, etc. custom button will not work (the effect I mean) when you change a (python) file of a map entirely, while when we do this in other game engines such as Panda3d, there is no problem.

**Comment by oskarmeyenburg** [comment]
I agree, it would be nice to have such features out of the box. But pygame is mostly a wrapper of SDL. This is maybe not what you are looking for, but there are other third-party packages providing such widgets, e.g. [pygame-menu](https://pygame-menu.readthedocs.io/en/latest/index.html) or [pygame_gui](https://pygame-gui.readthedocs.io/en/latest/index.html). Such unofficial versions also provide variety, which can certainly be an advantage.

**Comment by IGR2020** [comment]
This is very much what I was looking for! Thankyou!

**Comment by Fantastair** [comment]
@bambier I don't think it's scalable to implement a button component in isolation, we should first try to build a more scalable underlying display GUI framework and unified logic for event handling, on top of which we can write the button as an upper level component, the fact that I'm in the process of writing such an underlying framework helps us to focus our development on what to display rather than how to display it.

**Comment by Fantastair** [comment]
With such an underlying framework, text boxes, image boxes and input boxes can be implemented more easily.
In addition to graphical display and event handling, my underlying framework has a core element - keyframe animation, which is realized by setting a start state and a target state, which are then offset by a certain amount of time. The degree of offset is determined by a one-dimensional function (e.g., a specially transformed sine function can achieve silky-smooth movement animation), which borrows to some extent from the browser's animation logic.

**Comment by qhf566** [comment]
Who has a big factory, developed in the past 5 years, has not been operated in China, and has finished the development of the game, contact me 809996677
